File: backend/app/api/v1/categories.py
from typing import Optional
from fastapi import APIRouter, Depends, Query, status
from app.core.deps import require_admin
from app.schemas.category import (
    CategoryCreateRequest,
    CategoryStatusRequest,
    CategoryUpdateRequest,
)
from app.services.category_service import category_service
from app.utils.responses import success_response

router = APIRouter(tags=["Master Kategori Sampah"])


@router.get("/master/categories")
def list_categories(
    active_only: bool = Query(False, description="Tampilkan hanya kategori aktif"),
    is_active: Optional[bool] = Query(None, description="Alias untuk active_only"),
    has_active_price: bool = Query(False, description="Tampilkan hanya kategori yang memiliki harga aktif"),
    search: Optional[str] = Query(None, description="Pencarian nama kategori"),
    page: Optional[int] = Query(None, ge=1, description="Halaman"),
    page_size: Optional[int] = Query(None, ge=1, le=1000, description="Jumlah item per halaman"),
):
    """Daftar master kategori sampah beserta harga aktifnya (mendukung pagination opsional)."""
    effective_active_only = active_only or (is_active is True)
    if page is not None:
        effective_page_size = page_size or 20
        items, total_items = category_service.list_categories_paginated(
            active_only=effective_active_only,
            search=search,
            page=page,
            page_size=effective_page_size,
        )
        total_pages = (total_items + effective_page_size - 1) // effective_page_size if total_items > 0 else 0
        return success_response(
            data={
                "items": items,
                "pagination": {
                    "page": page,
                    "page_size": effective_page_size,
                    "total_items": total_items,
                    "total_pages": total_pages,
                },
            }
        )

    categories = category_service.list_categories(
        active_only=effective_active_only,
        search=search,
        has_active_price=has_active_price,
    )
    return success_response(data=categories)


@router.post("/admin/master/categories", status_code=status.HTTP_201_CREATED)
def create_category(data: CategoryCreateRequest, admin_user: dict = Depends(require_admin)):
    """Tambah kategori sampah baru (Admin)."""
    category = category_service.create_category(data, creator_id=admin_user["id"])
    return success_response(
        data=category,
        message="Kategori sampah berhasil ditambahkan.",
        status_code=status.HTTP_201_CREATED
    )


@router.put("/admin/master/categories/{category_id}")
def update_category(
    category_id: str, data: CategoryUpdateRequest, admin_user: dict = Depends(require_admin)
):
    """Ubah informasi kategori sampah (Admin)."""
    category = category_service.update_category(category_id, data, updater_id=admin_user["id"])
    return success_response(data=category, message="Kategori sampah berhasil diperbarui.")


@router.patch("/admin/master/categories/{category_id}/status")
def update_category_status(
    category_id: str, data: CategoryStatusRequest, admin_user: dict = Depends(require_admin)
):
    """Ubah status aktif/nonaktif kategori sampah (Admin)."""
    category = category_service.update_category_status(
        category_id, data.is_active, updater_id=admin_user["id"]
    )
    status_text = "diaktifkan" if data.is_active else "dinonaktifkan"
    return success_response(data=category, message=f"Kategori sampah berhasil {status_text}.")


@router.delete("/admin/master/categories/{category_id}")
def delete_category(category_id: str, admin_user: dict = Depends(require_admin)):
    """Hapus kategori sampah secara permanen (Admin)."""
    category_service.delete_category(category_id)
    return success_response(message="Kategori sampah berhasil dihapus.")

File: backend/src/routes/categories.ts
import { Hono } from 'hono';
import { AppEnv } from '../types/index.js';
import { categoryService } from '../services/categoryService.js';
import { priceService } from '../services/priceService.js';
import { authMiddleware } from '../middleware/auth.js';
import { requireAdmin } from '../middleware/roleGuard.js';
import { successResponse } from '../utils/response.js';
import { validateJson } from '../utils/validator.js';
import {
  categoryCreateSchema,
  categoryStatusSchema,
  categoryUpdateSchema,
} from '../schemas/category.schema.js';

// Public / Authenticated category router mounted at /api/v1/master/categories
export const categoriesRouter = new Hono<AppEnv>();

categoriesRouter.get('/', async (c) => {
  const query = c.req.query();
  const activeOnly =
    query.active_only === 'true' ||
    query.active_only === '1' ||
    query.is_active === 'true' ||
    query.is_active === '1';
  const result = await categoryService.listCategories({
    active_only: activeOnly,
    search: query.search,
    page: query.page ? parseInt(query.page, 10) : 1,
    page_size: query.page_size ? parseInt(query.page_size, 10) : 50,
  });
  return successResponse(c, result, 'Master kategori berhasil dimuat.');
});

categoriesRouter.get('/:id/active-price', async (c) => {
  const id = c.req.param('id');
  const result = await priceService.getActivePriceByCategory(id);
  return successResponse(c, result, 'Tarif aktif berhasil dimuat.');
});

categoriesRouter.get('/:id', async (c) => {
  const id = c.req.param('id');
  const result = await categoryService.getCategoryById(id);
  return successResponse(c, result, 'Detail kategori berhasil dimuat.');
});

// Admin category router mounted at /api/v1/admin/master/categories
export const adminCategoriesRouter = new Hono<AppEnv>();
adminCategoriesRouter.use('*', authMiddleware, requireAdmin);

adminCategoriesRouter.post('/', validateJson(categoryCreateSchema), async (c) => {
  const user = c.get('user');
  const data = c.req.valid('json' as any);
  const result = await categoryService.createCategory(data, user.id);
  return successResponse(c, result, 'Kategori sampah berhasil dibuat.', 201);
});

adminCategoriesRouter.put('/:id', validateJson(categoryUpdateSchema), async (c) => {
  const user = c.get('user');
  const id = c.req.param('id');
  const data = c.req.valid('json' as any);
  const result = await categoryService.updateCategory(id, data, user.id);
  return successResponse(c, result, 'Kategori sampah berhasil diperbarui.');
});

adminCategoriesRouter.patch('/:id/status', validateJson(categoryStatusSchema), async (c) => {
  const user = c.get('user');
  const id = c.req.param('id');
  const data = c.req.valid('json' as any);
  const result = await categoryService.updateCategoryStatus(id, data.is_active, user.id);
  return successResponse(c, result, 'Status kategori sampah berhasil diubah.');
});

adminCategoriesRouter.delete('/:id', async (c) => {
  const id = c.req.param('id');
  await categoryService.deleteCategory(id);
  return successResponse(c, null, 'Kategori sampah berhasil dihapus.');
});
